### PR TITLE
Add package file for xtb and py-xtb

### PR DESCRIPTION
**Pull request by awvwgk**
Add packages for xtb and py-xtb, also fixes simple-dftd3, dftd4 and mctc-lib packages to account for latest MesonPackage update.

### CODE REVIEW

**Review by tldahlgren — APPROVED** [review]
Confirmed the sha256 for the two new packages.  Rest LGTM.

**Comment by tldahlgren** [comment]
@adamjstewart Did you want to comment on the python aspects (e.g., `py-xtb`) of this PR?

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-xtb/package.py]: FYI. This will overwrite maintainers from the base package.  I'll let Adam comment on whether he likes this idea, but one option is to change this to `maintainers = PythonPackage.maintainers + ["awvwgk"]` so he can also be notified when changes are made.

**Review by tldahlgren — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (awvwgk) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: Thanks for the hint, I wasn't aware of this. It would be useful to have a function in `spack.package` to deal with this. Maybe something like this:

```diff
-maintainers = [
-    *PythonPackage.maintainers,
-    "user1",
-    "user2",
-]
+maintainer("user1")
+maintainer("user2")
```

---

Suggested in https://github.com/spack/spack/issues/34879.

Comment (awvwgk) [reply var/spack/repos/builtin/packages/py-xtb/package.py]:
```suggestion
    maintainers = [
        *PythonPackage.maintainers,
        "awvwgk",
    ]
```

File: var/spack/repos/builtin/packages/dftd4/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dftd4/package.py]: You can remove this and replace:
```python
depends_on("python@3.6:", when="+python")
```
with:
```python
extends("python", when="+python")
```
and it will have the same effect.

File: var/spack/repos/builtin/packages/mctc-lib/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mctc-lib/package.py:30]: @eli-schwartz what does this flag do and should we be using it?

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xtb/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xtb/package.py:19]: I don't particularly care about the default list of maintainers being overridden. I agree it would be nice to have a way to add maintainers instead of overwriting. If you want to submit a proposal for this I would be happy to review. `maintainers += ...` doesn't work does it?

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xtb/package.py]:
```suggestion
    depends_on("py-numpy", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-xtb/package.py]: Where does 0.11+ come from?

File: var/spack/repos/builtin/packages/simple-dftd3/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/simple-dftd3/package.py]: Same here, replace `depends_on` with `extends` and remove this function

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-xtb/package.py:19]: No, `maintainers += ..` results in `NameError: name 'maintainers' is not defined`.

The above also fails (on a simple test) with a `SyntaxError` on the asterisk.  

I've been recommending the following syntax:

`maintainers = PythonPackage.maintainers + ...`

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-xtb/package.py:19]: The "correct" way to do this is inside an `__init__()` function, probably. Class inheritance doesn't define this variable in advance, but looking it up via self.maintainers works.

Probably wisest to specify the behavior via a new dedicated `extra_maintainers` or something, yeah...

File: var/spack/repos/builtin/packages/mctc-lib/package.py
Comment (awvwgk) [reply var/spack/repos/builtin/packages/mctc-lib/package.py:30]: I only removed it because MesonPackage now applies it by default, so I don't have to enforce it on per package basis anymore.

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/mctc-lib/package.py:30]: This is a pretty useful flag, but you might want to set it in the default meson args from MesonPackage instead.

Basically, it's common for projects to vendor code for... various reasons. System integrators, such as linux distros, homebrew, spack, etc. rarely like this, and prefer using dependency packages instead.

Meson tries to solve this by making it *really* easy to use vendored code. So easy, in fact, that no one would dream of doing it in a manner other than the officially approved one. :D And that manner is, doing standard system dependency lookup plus an ini file defining an url to vendor it with, and what dependencies it contains. Then, Meson can guarantee via a global policy how to handle that.

- by default, it's:
  - searched as a system dependency (spack depends_on, in this case)
  - if not found, it downloads and vendors a copy
- if `--wrap-mode=nodownload` is used, then it will never download a copy, and prefer to error out if it must, but if one is already downloaded, it's eligible for use
- if `--wrap-mode=nofallback` is used, vendoring will NEVER be done, missing system dependencies are always errors
- if `--wrap-mode=forcefallback` is used, vendoring is forced, and system dependencies are totally ignored when vendoring is possible


Either nodownload or nofallback are reasonable defaults for spack to use.

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/mctc-lib/package.py:30]: Ah yes, just saw the new reply :D indeed, it's the default since 258edf7dac07710ea6c50316dc53e5154d9cb3b7 / https://github.com/spack/spack/pull/33717#pullrequestreview-1169543266 (which makes sense).

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (awvwgk) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: Numpy is actually not needed at build time.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (awvwgk) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: I have followed the development of meson-python  and it has been robustly usable for most of my projects since the 0.11 release. I can try to figure out if older versions will build successfully as well.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (awvwgk) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: This seems specific to the git-archive generated tarballs, here the sdist is already used from the GH release.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: This is invalid syntax (at the asterisk).

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: I would still prefer to use PyPI. We have much better support for finding new versions on PyPI (try `spack versions py-xtb` before and after) and PyPI tends to be more stable. See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#id2 for the correct syntax.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: It is if you run `spack install --test=root py-xtb`. If anyone can figure out how to add run-time deps to the path during a post-install test let me know, but for now it needs to be both.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: I would remove the constraint entirely and stick to what the developers require (no version constraint) unless you find evidence otherwise.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: Hmm, other packaging systems would have something like `type="test"` and this would be evaluated as part of the build dependencies but only if the install (or package creation) command specified to run tests.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: Spack also has a `type="test"`: https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types

We _could_ convert every `type=("build", "run")` to `type=("test", "run")` but it feels kinda silly. We could also drop the ability to test Python imports during installation and only allow them with `spack test` but I don't love that solution either. I wish Spack would just know how to handle this.

**Comment by awvwgk** [comment]
Thanks for the reviews, I hope I could address all comments.

File: var/spack/repos/builtin/packages/py-xtb/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-xtb/package.py]: Also, some build tools will check `PYTHONPATH` at build-time to make sure the dependencies are actually installed. So they do need to be build deps regardless.

File: var/spack/repos/builtin/packages/simple-dftd3/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/simple-dftd3/package.py]: This is a bit hard to read. Might be more legible as:
```suggestion
        url = "https://github.com/dftd3/simple-dftd3/releases/download/v{0}/s-dftd3-{0}{1}.tar.xz"
        source = ""
        if version >= Version("0.7"):
            source = "-source"
        return url.format(version, source)
```

Comment (awvwgk) [reply var/spack/repos/builtin/packages/simple-dftd3/package.py]: Fixed this upstream by naming the artifacts consistently

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
